packages/vite/src/node/utils.ts

```diff
@@ -189,8 +189,16 @@ const knownTsOutputRE = /\.(js|mjs|cjs|jsx)$/
 export const isTsRequest = (url: string) => knownTsRE.test(cleanUrl(url))
 export const isPossibleTsOutput = (url: string) =>
   knownTsOutputRE.test(cleanUrl(url))
-export const getTsSrcPath = (filename: string) =>
-  filename.replace(/\.([cm])?(js)(x?)(\?|$)/, '.$1ts$3')
+export const getPotentialTsSrcPaths = (filename: string) => {
+  // Special handling for filenames that end in `.js`, since they may actually
+  // refer to a `.ts` or `.tsx` file in TypeScript.
+  if (filename.endsWith('.js')) {
+    const extensionLessFile = filename.slice(0, -3)
+    return [`${extensionLessFile}.ts`, `${extensionLessFile}.tsx`]
+  }
```

Comment:
I believe the filename could potentially contain the `?` query parameter too (according to the regex), so checking `endsWith` could be risky. We would need to handle that, e.g. filename is `/foo/bar.js?bla`.

I'm also not sure if there's a case for `.cjs` and `.mjs` to map to `.ctsx` and `.mtsx`. But this PR would be in the next meeting to be discussed.

Reply:
Cool, will think about how to better handle cases with `?` after the extension. Also still need to add tests for `resolve`. Let me know how the meeting goes. I'll be happy to update this PR based on any decisions made there.